### ISSUE TITLE
axis argument for FFT ops (tf.signal.fft, tf.signal.fft2d, etc.)

### ISSUE THREAD

**Issue by hsgkim** [tensorflow/tensorflow#23156]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 1.11
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
`tf.spectral.fft3d` as of now is restricted to apply FFT to the last 3 dimensions of the input, while the majority of the operations, e.g. `tf.nn.conv3d`, handle data in NDHWC format by default. In order to circumvent this limitation, users have to permute (transpose) the data just for the FFT operation, then permute them back for other operations that follow thereafter. This not only makes the code less readable, but more importantly incurs extra overhead.

**Will this change the current api? How?**
This will add another argument `data_format` to `tf.spectral.fft3d` which lets users decide which dimensions should be used for FFT. `data_format` should be either `"NDHWC"` or `"NCDHW"`, and defaults to `"NCDHW"` for backward compatibility. If `"NDHWC"` is chosen as the `data_format`, the second, third and fourth dimensions will be used instead. 

**Who will benefit with this feature?**
Any users using `fft3d` are expected to benefit from this, especially when they use FFT with other convolutional layers that accept inputs in NDHWC format.

**Any Other info.**
Other operations such as `tf.spectral.fft2d` may also be modified for consistency.

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce
Mobile device

**Comment by hsgkim** [comment]
Have I written custom code N/A
OS Platform and Distribution N/A
TensorFlow installed from N/A
Bazel version N/A
CUDA/cuDNN version N/A
GPU model and memory N/A
Exact command to reproduce N/A
Mobile device N/A

**Comment by wt-huang** [comment]
@hsgkim Agreed that this is not efficient let alone error prone. 
tf.spectral.fft3d should cover both "NDHWC" or "NCDHW". Feel free to contribute a PR for this.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @wt-huang: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by rryan** [comment]
`tf.signal.fft` (and its variants) are designed for `numpy.fft` and `scipy.signal` compatibility. The `data_format` layouts for TensorFlow's convolution operators only make sense in the context of a convolution, since there is no channel dimension in an FFT, so I do not think we should do this. Instead, I would prefer that `tf.signal.fft` had an axis or axes argument, as `numpy.fft.fft` and `numpy.fft.fft2` do.

Changed the issue description. I don't think I'll have time to work on this soon, so I'm adding the contributions welcome tag.

**Comment by NEGU93** [comment]
+1 for this. I would like `tf.signal.fft2d` to have the `axes` parameters as [`numpy.fft.fft2`](https://numpy.org/doc/stable/reference/generated/numpy.fft.fft2.html) for example (and all analogous examples). This will also enable the creation of `tf.signal.fftnd`

**Comment by shariq-audiofocus** [comment]
would love this feature too! An axis argument would be helpful when using stereo audio, where the channel is conventionally the inner most dimension, and the axis to compute the FFT on, i.e. time/sample dimension, is the second inner most dimension.

**Comment by huseinzol05** [comment]
+1, need this.

**Comment by Yaxit** [comment]
+1, would be super useful

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
